### ISSUE TITLE
Merge multiple config files

### ISSUE THREAD

**Issue by lyz-code** [dbatten5/maison#61]
Hi again @dbatten5, we've been discussing in [an autoimport issue](https://github.com/lyz-code/autoimport/issues/169) that it will be nice to have a global config somewhere in `~/.config/autoimport/config.toml` besides the proyect `pyproject.toml`. That means that we need to take the global config as the base and update it with the values of the project config.  @HoverHell has implemented how to do that in [this PR](https://github.com/lyz-code/autoimport/pull/172#pullrequestreview-828254747), but I feel that as `maison` already supports a list of `source_files` it would make sense to have that config merging code in `maison` and not in `autoimport`. That way I could reuse it for the rest of my projects, such as `yamlfix` :P.

What do you think?

**Comment by dbatten5** [comment]
hi @lyz-code 👋. i remember coming across this bit of `mypy` [documentation](https://mypy.readthedocs.io/en/stable/config_file.html) when developing this package initially which put me off implementing merging of configs in the first place:

![image](https://user-images.githubusercontent.com/17271241/145549480-069bc74f-4caa-414c-99f3-7c518003bbd0.png)

however, conceptually this idea does make sense in my head and it seems reasonable that `maison` should offer it as an option. i'll take a look into it although i might not have capacity to release it before a week or two

**Comment by lyz-code** [comment]
Thanks for considering it :), and no rush at all. Maybe @HoverHell opens a PR with the changes it's made in `autoimport`

**Comment by dbatten5** [comment]
yeah a pr would be most welcome @HoverHell if you'd like to open one here

**Comment by HoverHell** [comment]
> yeah a pr would be most welcome @HoverHell <https://github.com/HoverHell>
> if you'd like to open one here.
>
What do you expect the interface (API) to look like? Combining both "look
up in order" and "merge in order". A sequence (merge) of sequence (find
first) of paths?

As for ambiguity, I would argue `autoimport` is not the worst case, as it
is an IDE-like interface tool, rather than completely automated CI tool (as
far as I understand). Of course, with an option to disable merging that
should be even less of a problem.

**Comment by dbatten5** [comment]
how about a `merge_configs` boolean in the constructor of `ProjectConfig`? if `True` then all configs listed in `source_files` that are found on the filesystem are merged from right to left, if `False` then the behaviour is as it currently is (the first config found in `source_files` is used)

**Comment by Jasha10** [comment]
Hi @dbatten5,
Let me introduce myself -- I'm Jasha, a downstream user of `maison` via `autoimport`.
Thank you for maintaining `maison`!

> ... are merged from right to left ...

I wanted to ask, by right-to-left, do you mean that the rightmost file would take the highest precedence? This seems most intuitive to me as it mirrors the right-precedence behavior of the `dict.update` and the `dict.__or__` methods.

**Comment by dbatten5** [comment]
hi @Jasha10, thanks for getting touch and glad you're finding some use in `maison`
yeah as you say i think the most intuitive behaviour would be that rightmost config sources take precedence. i think with some good documentation it could be made clear that the most important config should come last in the list. 

so for something like:
```python
["~/.config/acme/config.toml", "~/.acme.toml", ..., "acme.ini", "pyproject.toml"]
```
`"~/.config/acme/config.toml"` could set some base global configs which could be refined/overridden in a `"/path/to/project/pyproject.toml"` `[acme.tool]` section

does that make sense?

**Comment by Jasha10** [comment]
Makes sense to me!

**Comment by lyz-code** [comment]
Investigating [PDM](https://github.com/pdm-project/pdm/) I've found out that they do something similar to our wishes, maybe we can use [their code](https://github.com/pdm-project/pdm/blob/main/pdm/project/config.py) as a starting point

**Comment by dbatten5** [comment]
ah nice. i've taken the first steps towards this by adding support for absolute paths to configs in https://github.com/dbatten5/maison/pull/71. hopefully can find some time in the next few weeks to do the rest. of course if anyone else would like to tackle it they'd be most welcome

**Comment by dbatten5** [comment]
i've done a POC here: https://github.com/dbatten5/maison/pull/72
do you think this would suit your needs? the only thing that's changed in the public api is a `merge_configs` boolean that can be passed in the constructor for `ProjectConfig`. if `True` then all existing configs will be merged right to left.
if people are happy with it i can start adding tests etc.
@Jasha10 @lyz-code @HoverHell

**Comment by lyz-code** [comment]
It looks perfect to me @dbatten5 thank you so much :)

Completely offtopic, but maybe it could interest you. I don't know if you know [pydantic's settings management](https://pydantic-docs.helpmanual.io/usage/settings/), or the yaml implementation [goodconf](https://github.com/lincolnloop/goodconf/), maybe you can get ideas or reuse code. Sorry for not pointing it out earlier.

**Comment by dbatten5** [comment]
i've tidied up the pr with tests etc. if you'd like to review you'd be very welcome to, otherwise will get it merged in later today and get a new release out with the merge configs options.

yeah big fan of the pydantic's settings mgmt. goodconf looks nice too, will take a look at that

**Comment by dbatten5** [comment]
this is available in 1.3.0

**Comment by lyz-code** [comment]
Awesome! thank you so much :)
